File: aat-android/src/main/java/ch/bailu/aat/map/mapsforge/MapsForgeView.java
package ch.bailu.aat.map.mapsforge;

import android.graphics.Canvas;

import org.mapsforge.core.model.LatLong;
import org.mapsforge.map.model.common.Observer;

import ch.bailu.aat.map.MapDensity;
import ch.bailu.aat_lib.map.layer.MapPositionLayer;
import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat_lib.dispatcher.DispatcherInterface;

public class MapsForgeView extends MapsForgeViewBase {

    private final MapsForgeForeground foreground;
    private final MapsForgeTileLayerStackConfigured stack;
    private final MapPositionLayer pos;

    public MapsForgeView(ServiceContext sc, DispatcherInterface dispatcher, String key) {
        super(sc, key, new MapDensity(sc.getContext()));

        stack = new MapsForgeTileLayerStackConfigured.All(this);
        add(stack, stack);

        // Depends on zoom limits (setItem by TileLayerStack)
        pos = new MapPositionLayer(getMContext(), new Storage(sc.getContext()), dispatcher);
        add(pos);

        foreground = new MapsForgeForeground(this,
                getMContext(),
                new MapDensity(sc.getContext()),
                getLayers());

        setClickable(true);

        getModel().mapViewPosition.addObserver(new Observer() {
            private LatLong center = getModel().mapViewPosition.getCenter();

            @Override
            public void onChange() {
                LatLong newCenter = getModel().mapViewPosition.getCenter();

                if (newCenter != null && newCenter.equals(center) == false) {
                    center = newCenter;
                    pos.onMapCenterChanged(center);
                }
            }
        });

    }


    @Override
    public void reDownloadTiles() {
        stack.reDownloadTiles();
    }


    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        foreground.dispatchDraw(canvas);
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/map/mapsforge/MapsForgeView.kt
package ch.bailu.aat.map.mapsforge

import android.content.Context
import android.graphics.Canvas
import ch.bailu.aat.map.MapDensity
import ch.bailu.aat.preferences.Storage
import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.dispatcher.DispatcherInterface
import ch.bailu.aat_lib.map.layer.MapPositionLayer
import ch.bailu.aat_lib.map.tile.MapsForgeTileLayerStackConfigured
import org.mapsforge.map.model.common.Observer

class MapsForgeView(
    context: Context,
    appContext: AppContext,
    dispatcher: DispatcherInterface,
    key: String
) : MapsForgeViewBase(appContext, context, key, MapDensity(context)) {

    private val stack = MapsForgeTileLayerStackConfigured.All(this, appContext)
    private val pos = MapPositionLayer(getMContext(), Storage(context), dispatcher)
    private val services = appContext.services
    private val foreground = MapsForgeForeground(appContext, this, getMContext(), MapDensity(context), layers)

    init {
        add(stack, stack)
        add(pos)

        isClickable = true
        model.mapViewPosition.addObserver(object : Observer {
            private var center = model.mapViewPosition.center
            override fun onChange() {
                val newCenter = model.mapViewPosition.center
                if (newCenter != null && newCenter != center) {
                    center = newCenter
                    pos.onMapCenterChanged(center)
                }
            }
        })
    }

    override fun reDownloadTiles() {
        stack.reDownloadTiles()
    }

    public override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        foreground.dispatchDraw(services, canvas)
    }
}
